mpf/devices/blinkenlight.py

```diff
@@ -0,0 +1,131 @@
+from mpf.core.delays import DelayManager
+from mpf.core.device_monitor import DeviceMonitor
+from mpf.exceptions.config_file_error import ConfigFileError
+
+@DeviceMonitor("num_colors", "light")
+class Blinkenlight(SystemWideDevice):
+
+    """A light that alternates between several different colors."""
+
+    config_section = 'blinkenlights'
+    collection = 'blinkenlights'
+    class_label = 'blinkenlight'
+
+    def __init__(self, machine, name):
+        """Initialise blinkenlight."""
+        super().__init__(machine, name)
+
+        self._colors = []
+        self.delay = DelayManager(machine)
+        self._num_colors = 0
+
+    def load_config(self, config: dict):
+        super().load_config(config)
+
+        self._color_duration = self.config['color_duration']
+        self._cycle_duration = self.config['cycle_duration']
+        if (self._color_duration is None and self._cycle_duration is None) or (self._color_duration is not None and self._cycle_duration is not None):
+            self._blinkenlight_validation_error("Either color_duration or cycle_duration must be specified, but not both.", 1)
+
+    def _blinkenlight_validation_error(self, msg, error_code) -> "NoReturn":  # pragma: no cover
+        raise ConfigFileError('"{}" >> {}'.format(self.name, msg), error_code, "blinkenlight", self.name)
+
+    @property
+    def num_colors(self):
+        # I would prefer this to be a calculated value (i.e., just return len(self._colors)), but then there's no setter involved so things like mpf monitor won't know when this value gets updated.
+        return self._num_colors
+
+    @num_colors.setter
+    def num_colors(self, newValue):
+        self._num_colors = newValue
+        # if the number of colors is 0, remove us from the light stack
+        if self._num_colors == 0:
+            self.light.remove_from_stack_by_key(self._blinkenlight_key)
+
+    @property
+    def light(self):
+        return self.config['light']
+
+    @property
+    def num_colors_in_cycle(self):
+        # similar to num_colors, but adds 1 for the "off" color if there is one between cycles
+        return self.num_colors + (1 if self._off_between_cycles() else 0)
+
+
+    @property
+    def _blinkenlight_key(self):
+        # this is the key that's passed to the light as the color's key
+        # so that we can remove this blinkenlight's color from the light stack
+        # when all the blinkenlight's colors are removed.
+        # This prevents the light from staying in the off state if the blinkenlight
+        # was on top of the priority list.
+        return 'blinkenlight_{}'.format(self.name)
+
+    def add_color(self, color, key):
+        # only add this color if the key does not already exist
+        if len([x for x in self._colors if x[1] == key]) < 1:
+            self._colors.append((color, key))
+            self.num_colors += 1
+            self.info_log('Color {} with key {} added'.format(color, key))
+
```

Comment:
I would order the list by priority here

Reply:
I order the list in _restart, since add_color, remove_color, and remove_all_colors calls _restart.